File: tts.go
package main

import (
	"crypto/tls"
	"encoding/json"
	"log"
	"net/http"
	"strings"
	"sync"
	"time"

	"github.com/google/uuid"
	"github.com/gorilla/websocket"
)

func ttsManagerLoop() {
	var conn *websocket.Conn
	var wg sync.WaitGroup
	var currentTaskID string
	var localSessionID string
	taskStartedSignal := make(chan struct{}, 1)
	var firstPacketReceived bool

	isCanceled := func() bool {
		ctxMutex.Lock()
		defer ctxMutex.Unlock()
		return sessionCtx.Err() != nil
	}

	receiveLoop := func(c *websocket.Conn) {
		defer wg.Done()
		defer func() {
			if !isCanceled() {
				audioPcmChan <- []byte{}
			}
		}()
		for {
			if isCanceled() {
				return
			}
			msgType, msg, err := c.ReadMessage()
			if err != nil {
				return
			}
			if msgType == websocket.BinaryMessage {
				if !firstPacketReceived {
					tsFirstAudio = time.Now()
					firstPacketReceived = true
					log.Printf("TTS 首包: %v", tsFirstAudio.Sub(tsTtsStart))
				}
				if !isCanceled() {
					audioPcmChan <- msg
				}
				continue
			}
			var resp map[string]interface{}
			if err := json.Unmarshal(msg, &resp); err != nil {
				continue
			}
			header, _ := resp["header"].(map[string]interface{})
			event := header["event"].(string)
			if event == "task-started" {
				select {
				case taskStartedSignal <- struct{}{}:
				default:
				}
			}
			if event == "task-finished" || event == "task-failed" {
				return
			}
		}
	}

	for {
		msg, ok := <-ttsManagerChan
		if !ok {
			return
		}

		sessionIDMutex.Lock()
		globalID := currentSessionID
		sessionIDMutex.Unlock()

		if localSessionID != globalID {
			if conn != nil {
				conn.Close()
				conn = nil
			}
			localSessionID = globalID
		}

		if isCanceled() {
			if conn != nil {
				conn.Close()
				conn = nil
			}
			continue
		}

		if msg == "[[END]]" {
			if conn != nil {
				conn.WriteJSON(map[string]interface{}{
					"header":  map[string]interface{}{"task_id": currentTaskID, "action": "finish-task", "streaming": "duplex"},
					"payload": map[string]interface{}{"input": map[string]interface{}{}},
				})
				wg.Wait()
				conn.Close()
				conn = nil
			}
			continue
		}

		if strings.TrimSpace(msg) != "" {
			if conn == nil {
				dialer := websocket.Dialer{TLSClientConfig: &tls.Config{InsecureSkipVerify: true}}
				headers := http.Header{}
				headers.Add("Authorization", "Bearer "+dashAPIKey)
				c, _, err := dialer.Dial(ttsWsURL, headers)
				if err != nil {
					continue
				}
				conn = c
				currentTaskID = uuid.New().String()
				firstPacketReceived = false
				tsTtsStart = time.Now()
				wg.Add(1)
				go receiveLoop(conn)
				conn.WriteJSON(map[string]interface{}{
					"header": map[string]interface{}{"task_id": currentTaskID, "action": "run-task", "streaming": "duplex"},
					"payload": map[string]interface{}{
						"task_group": "audio", "task": "tts", "function": "SpeechSynthesizer",
						"model":      ttsModel,
						"parameters": map[string]interface{}{"text_type": "PlainText", "voice": ttsVoice, "format": "pcm", "sample_rate": ttsSampleRate, "volume": ttsVolume, "enable_ssml": false},
						"input":      map[string]interface{}{},
					},
				})
				select {
				case <-taskStartedSignal:
					time.Sleep(50 * time.Millisecond)
				case <-time.After(5 * time.Second):
					conn.Close()
					conn = nil
					continue
				}
			}
			conn.WriteJSON(map[string]interface{}{
				"header":  map[string]interface{}{"task_id": currentTaskID, "action": "continue-task", "streaming": "duplex"},
				"payload": map[string]interface{}{"input": map[string]interface{}{"text": msg}},
			})
			time.Sleep(50 * time.Millisecond)
		}
	}
}

File: py_ai_box/tts.py
import json
import threading
import time
import uuid

from . import state
from . import config_runtime
from .websocket_client import WebSocketClient


def tts_manager_loop() -> None:
    conn = None
    current_task_id = None
    local_session_id = None
    recv_thread = None
    task_started_event = threading.Event()
    first_packet_received = False

    def is_canceled(ev):
        return ev.is_set()

    def receive_loop(ws: WebSocketClient, ev):
        nonlocal first_packet_received
        try:
            while True:
                if is_canceled(ev):
                    return
                try:
                    msg_type, payload = ws.recv()
                except RuntimeError:
                    return
                if msg_type == "binary":
                    if not first_packet_received:
                        state.ts_first_audio = time.time()
                        first_packet_received = True
                        if state.ts_tts_start is not None:
                            print(f"TTS 首包: {state.ts_first_audio - state.ts_tts_start:.3f}s")
                    if not is_canceled(ev):
                        state.audio_pcm_queue.put(payload)
                    continue
                if msg_type != "text":
                    continue
                try:
                    resp = json.loads(payload)
                except Exception:
                    continue
                header = resp.get("header", {})
                event = header.get("event")
                if event == "task-started":
                    task_started_event.set()
                if event in ("task-finished", "task-failed"):
                    return
        finally:
            if not is_canceled(ev):
                state.audio_pcm_queue.put(b"")

    while True:
        if state.shutdown_event.is_set():
            if conn is not None:
                conn.close()
            return
        msg = state.tts_manager_queue.get()
        with state.session_id_lock:
            global_id = state.current_session_id

        if local_session_id != global_id:
            if conn is not None:
                conn.close()
                conn = None
            local_session_id = global_id

        if is_canceled(state.session_cancel_event):
            if conn is not None:
                conn.close()
                conn = None
            continue

        if msg == "[[END]]":
            if conn is not None:
                conn.send_text(json.dumps({
                    "header": {"task_id": current_task_id, "action": "finish-task", "streaming": "duplex"},
                    "payload": {"input": {}},
                }, ensure_ascii=False))
                if recv_thread is not None:
                    recv_thread.join(timeout=5)
                conn.close()
                conn = None
            continue

        if msg.strip():
            if conn is None:
                conn = WebSocketClient(config_runtime.tts_ws_url, headers={"Authorization": f"Bearer {config_runtime.dash_api_key}"})
                try:
                    conn.connect()
                except Exception:
                    conn = None
                    continue
                current_task_id = uuid.uuid4().hex
                first_packet_received = False
                state.ts_tts_start = time.time()
                task_started_event.clear()
                recv_thread = threading.Thread(
                    target=receive_loop,
                    args=(conn, state.session_cancel_event),
                    daemon=True,
                )
                recv_thread.start()
                conn.send_text(json.dumps({
                    "header": {"task_id": current_task_id, "action": "run-task", "streaming": "duplex"},
                    "payload": {
                        "task_group": "audio",
                        "task": "tts",
                        "function": "SpeechSynthesizer",
                        "model": config_runtime.tts_model,
                        "parameters": {
                            "text_type": "PlainText",
                            "voice": config_runtime.tts_voice,
                            "format": "pcm",
                            "sample_rate": config_runtime.tts_sample_rate,
                            "volume": config_runtime.tts_volume,
                            "enable_ssml": False,
                        },
                        "input": {},
                    },
                }, ensure_ascii=False))
                if not task_started_event.wait(timeout=5):
                    conn.close()
                    conn = None
                    continue
                time.sleep(0.05)

            conn.send_text(json.dumps({
                "header": {"task_id": current_task_id, "action": "continue-task", "streaming": "duplex"},
                "payload": {"input": {"text": msg}},
            }, ensure_ascii=False))
            time.sleep(0.05)
